src/librustc_typeck/check/mod.rs

```diff
@@ -5010,14 +5015,49 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                 path_segs.push(PathSeg(generics_def_id, last));
             }
 
-            // Case 2. Reference to a top-level value.
+            // Case 2. Reference to a variant constructor.
+            Def::VariantCtor(def_id, ..) => {
```

Comment:
It would be nice to have tests for various configurations of generic arguments provided to both enums and variants, for both "variant types" and variant constructors.

Reply:
Rough test plan:
```rust
enum Enum<T> { Variant }
type Alias<T> = Enum<T>;
type AliasFixed = Enum<u8>;

// All combinations (2*2*3)
// For successful cases some check should ensure that the resulting type after substitutions is indeed what we are expecting
Enum</*params | no params*/>::Variant</*params | no params*/>
Alias</*params | no params*/>::Variant</*params | no params*/>
AliasFixed</*params | no params*/>::Variant</*params | no params*/>
```

Reply:
Oh, also same for `Self` as a fixed alias.

Reply:
Okay, thanks.

Reply:
What do you mean "fixed alias"?

Reply:
`Self` behaving like `AliasFixed`.